### PR TITLE
Web app for managing InferenceServices

### LINKED ISSUE #1327: Web app for managing InferenceServices
/kind feature

**Describe the solution you'd like**
Have a distinct web app for managing the life-cycle of v1beta1 `InferenceServices` and allow the users to perform CRUD operations on them without the need for `kubectl`.

This app should follow the same guidelines we have for Kubeflow web apps like using `SubjectAccessReviews` to perform authorization checks https://github.com/kubeflow/community/pull/334 as well as be able to be deployed as a standalone component https://github.com/kubeflow/kubeflow/issues/5566.

**Anything else you would like to add:**
Ideally this web app should be able to share some common code and UX with the other web apps we have in Kubeflow, like Notebooks, Katib and Volumes web apps. This will make it easier to maintain the apps, since we won't be duplicating the effort to fix common issues and add new features.

### PR DESCRIPTION
**Pull request by kimwnasptd**
## What this PR does / why we need it
We (Arrikto) have created a web app for managing `InferenceService` CRs as part of our [KubeCon 2020 US presentation](https://youtu.be/VDINH5WkBhA?t=2504), which we also [demoed](https://youtu.be/OKqx3IS2_G4?list=PLmzRWLV1CK_ypvsQu10SGRmhf2S7mbYL5&t=1219) to the community. This web app has been available in MiniKF, for both GCP and AWS, since November, with significant usage numbers, reported by the two cloud providers. 

Driven by the hours of usage and the positive feedback we've received from users, we decided to contribute this web app to the upstream KFServing repo, as we believe it will very much improve the experience of end users. This web app is built to manage [v1beta1](https://github.com/kubeflow/kfserving/blob/master/install/v0.5.0-rc2/kfserving.yaml) `InferenceService` CRs, which will be available in the 1.3 release.

I'd like to also note that we are committed to keep up with the latest work and updates from the Serving WG, as well as feedback from end users, in order to ensure that the app stays up to date and rigorously maintained.

**Which issue(s) this PR fixes** :
first work for #1327
We should not close this issue, but rather keep it open to discuss future design decisions and link other issues to it.

## Special notes for your review

We understand that the Serving WG, and the current contributors of this repo, might not have the resources to help maintain or review this web app as it evolves, since there might not be a lot of people with UI expertise in the WG. To tackle this, the following people from our side commit to maintain it going forward: @StefanoFioravanzo @elikatsis @tasos-ale.

So by having the engineering effort covered, we believe that the only crucial part for maintaining this app will be:
1. our ability to effectively communicate our state 
2. **discuss** and **agree** on what we need, via GitHub issues, before implementing anything

This way the development process will be tightly integrated with the community and people will always be able to know the state of this component, even if they are not familiar with the underlying technologies.

## Release note
This web app is compliant with the latest [v1beta1](https://github.com/kubeflow/kfserving/blob/master/install/v0.5.0-rc2/kfserving.yaml) API and expects this version `InferenceService` objects.

At this point we would like to also present some screenshots so that you can also have a quick look of the app

Screenshot_2021-01-29 Kubeflow Central Dashboard

Screenshot_2021-01-29 Kubeflow Central Dashboard(2)

<details><summary>More pictures</summary>
<p>

**An overview of the CR and its live status**
Screenshot_2021-01-29 Kubeflow Central Dashboard(1)

**Details of the CR provided in a user-friendly way**
Screenshot_2021-01-29 Kubeflow Central Dashboard(3)

**Metrics of the InferenceService**
Screenshot_2021-01-29 Kubeflow Central Dashboard

**Viewing the entire YAML of the CR**
Screenshot_2021-01-29 Kubeflow Central Dashboard(1)

**And last but not least the form for creating a new InferenceService**
Screenshot_2021-01-29 Kubeflow Central Dashboard(2)
</p>
</details>

### CODE REVIEW

**Comment by pugangxa** [comment]
Great contribution!!! 
Some questions:
- So this must be component for Kubeflow GUI but cannot be deployed as dedicated kfserving GUI right?
- There's some component in Kubeflow are implemented in react and the backend is implemented by Golang(like pipeline), is there a reason why we use angular and Flask here?
- Is there any testing which we can integrated with the CI/CD system, I think the backend will get data from k8s directly(while in pipeline we just refer the go struct inside the same repo) so to find some compatibility issue we'd better have some testing.

**Comment by kimwnasptd** [comment]
Hi there @pugangxa 
> So this must be component for Kubeflow GUI but cannot be deployed as dedicated kfserving GUI right?

The plan for this GUI is to be deployed with standalone KFServing as well. We deliberately did not make the app deploy-able as standalone just yet, because we want to first coordinate and agree with the community on what standalone mean exactly.

Since this is a common issue for other web apps as well, like Notebooks, Katib, Tensorboards, etc we have already opened the issue https://github.com/kubeflow/kubeflow/issues/5566 in which we discuss our assumptions and understanding.

If the Serving WG agrees with our understanding in that issue as well, then as a subsequent step, we will include this functionality too.

> There's some component in Kubeflow are implemented in react and the backend is implemented by Golang(like pipeline), is there a reason why we use angular and Flask here?

We are currently moving to have the Notebooks, Volumes, Tensorboards and Katib UI ( frontend https://github.com/kubeflow/katib/issues/1421 ) web apps, use these same technologies.

All of these apps also share a lot of common code, which allows us to easily fix issues uniformly and reduce the maintenance effort and corresponding bugs. https://github.com/kubeflow/kubeflow/pull/5164 https://github.com/kubeflow/kubeflow/pull/5252

> Is there any testing which we can integrated with the CI/CD system, I think the backend will get data from k8s directly(while in pipeline we just refer the go struct inside the same repo) so to find some compatibility issue we'd better have some testing.

Correct, the backend fetches data directly from the k8s api server.

We plan on making a subsequent PR right after this one to extend the current CI to run some unitttests, as a first step. We will include scripts that generate Argo Workflows, via python and not ksonnet which is what's currently used, for these tests. The tests will also only run if the source code of the app is run, in order to not utilize any unnecessary resources.

Then the next step will be to add some e2e tests, again by extending the prow jobs, that will ensure we catch early compatibility issues.

**Comment by pugangxa** [comment]
@kimwnasptd Thanks for the detailed information, it's fine to me now. @yuzisun , I think kfserving is relatively independent from kubeflow and if there's dedicated GUI it will be great, what do you think?

**Review by nlin-kobold — APPROVED** [review]
👍 really nice feature

**Comment by davidyuyuan** [comment]
Thanks for the great effort to enhance the usability of the platform. A nice UI saves a lot of effort for the end users, and always highly appreciated.

**Comment by LeonardAukea** [comment]
Great Initiative! This looks awesome.

**Comment by pvaneck** [comment]
Hey, thanks for submitting this! I tried it out locally, and it seems to work great. However, I am not sure on how to get the Metrics tab going. I might be missing something. Do you have instructions on how Grafana or the web app should be configured to allow the web app to show InferenceService metrics?

**Comment by kimwnasptd** [comment]
For everyone who is following the progress of the PR we've also uploaded a short video to get some hands on view of the new UI https://www.youtube.com/watch?v=L0cTPM4I8CU

**Comment by kimwnasptd** [comment]
Hi there @pvaneck, thank you for trying this out and for the feedback! I discuss about the metrics tab in my demo video, mentioned in the comment above, in case this is helpful. As you've mentioned you'll need to have Grafana exposed to be able to see this tab.

I used Grafana from the [Metrics Plugin](https://github.com/knative/serving/releases/v0.18.0/) of Knative, plus some small configurations to expose it outside the cluster. I'd like to keep the discussion of installing/exposing Grafana outside the scope of this PR, since this is a complete conversation of each own.

I'd be happy to help you with Grafan though. Would it make sense to create a thread in Kubeflow's slack and continue this discussion there?

**Comment by kimwnasptd** [comment]
Since this is a huge PR with a lot of new functionalities we've prepared a design doc to help the reviewer process. It aims to 
1. expose the design decisions behind the web app
2. explain the functionalities of the web app
3. expose some of its weak points and our thoughts on next steps and features

https://docs.google.com/document/d/1MaCqHeqd7NiP1nseWkzqqAntRr3PiEH7k_7WnKjQSPU/edit?usp=sharing

cc @kubeflow/wg-serving-leads

**Comment by kimwnasptd** [comment]
Hi everyone, I think people have had enough time to review what I am proposing, are there any concerns that need to be addressed?

**Review by pvaneck — COMMENTED** [review]
Hey, sorry for getting back late. Thanks for putting together that doc and video, as that really helped shed light on the app's current states and limitations, as well as helping me understand the metrics side of things. Good to see that the metrics tab is hidden if the Grafana endpoint can't be reached, and also nice to see the planned support for standalone mode.

All in all, I think the PR would be an awesome addition. Do you plan on adding deployment/service YAMLs and the configuration necessary to integrate this into the Kubeflow dashboard? (I'm assuming I'd have to update the centraldashboard config to add this app as a menu item too)

**Comment by kimwnasptd** [comment]
> All in all, I think the PR would be an awesome addition.

Thank your very much @pvaneck!

> Do you plan on adding deployment/service YAMLs and the configuration necessary to integrate this into the Kubeflow dashboard?

Yes, I do. I have the manifests almost ready, since we are using them for MiniKF. I was thinking of pushing them as a PR on their own so that we can review and iterate on them separately.

I'm also planning on making the required PR to integrate this with Kubeflow's dashboard, but I'd prefer if we'd take it step by step. I believe our next goal for this effort should be to integrate this app to kfserving's CI/CD so that we can have images for the app. Then, once we have images, I can make a PR to add this app in Kubeflow's dashboard.

File: web-app/frontend/src/app/shared/utils.ts
Comment (kimwnasptd) [reply web-app/frontend/src/app/shared/utils.ts:139]: Good catch, will push a commit that addresses this

File: web-app/backend/apps/common/versions.py
Comment (kimwnasptd) [reply web-app/backend/apps/common/versions.py:20]: Indeed, I'll add it

File: web-app/frontend/src/app/shared/utils.ts
Comment (kimwnasptd) [reply web-app/frontend/src/app/shared/utils.ts]: pushed a commit for this. All instances are `lightgbm` now

File: web-app/backend/apps/common/versions.py
Comment (kimwnasptd) [reply web-app/backend/apps/common/versions.py]: Fixed the typo and made the `v1beta2` to be `v1beta1`

**Comment by kimwnasptd** [comment]
@pvaneck do you have any insights as to why the `Python package / build` test fails? From the logs I see
```
xgbserver/xgbserver/test_xgboost_model_repository.py:32
  Unknown pytest.mark.asyncio - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/mark.html

xgbserver/xgbserver/test_model.py::test_model
  Initializing DMatrix from List is deprecated.

-- Docs: https://docs.pytest.org/en/stable/warnings.html
=========================== short test summary info ============================
FAILED alibiexplainer/tests/test_anchor_images.py::test_anchor_images - reque...
================= 1 failed, 133 passed, 919 warnings in 51.70s =================
Error: Process completed with exit code 1.
```
I've also rebased on top of the latest master.

**Comment by kimwnasptd** [comment]
Αnd lastly what are the next steps i could to to assist the review? What could be the next steps here?

@yuzisun I remember from the meeting that you would give a look at the design-doc/PR. Just a kind ping on this. Is this still the plan? Can I do anything to help you any further?

**Comment by theofpa** [comment]
Image-net changed their API today, [ref](http://image-net.org/update-mar-11-2021.php). The test fails because alibi downloads the bounding boxes from the image-net API ([ref](https://github.com/SeldonIO/alibi/blob/76e6192b6d78848dd47c11ba6f6348ca94c424c6/alibi/datasets.py#L95)) which is no longer available. The xml files with the bounding boxes coordinates are available in [kaggle](https://www.kaggle.com/c/imagenet-object-localization-challenge/data).

**Comment by animeshsingh** [comment]
@kimwnasptd thanks for this - looks great and a welcome addition!
I think standalone mode would make more sense to come in KFServing code-base. When run in conjunction with rest of Kubeflow, it can then be enabled for multi-user et all via configuration settings?

Also one question I have is that if all the components use the same backend infra for web ui, wouldn't it make sense to have a common dashboard project hold all the UI code for different Kubeflow components?

Additionally did you ensure the monitoring can work with monitoring setup here https://github.com/kubeflow/kfserving/blob/master/docs/samples/metrics-and-monitoring/README.md#install-prometheus

If not, please suggest changes which will need to go in monitoring setup. 

@pvaneck @sriumcp @yuzisun @Tomcli

**Comment by kimwnasptd** [comment]
> @kimwnasptd thanks for this - looks great and a welcome addition!
> I think standalone mode would make more sense to come in KFServing code-base. When run in conjunction with rest of Kubeflow, it can then be enabled for multi-user et all via configuration settings?

Thank you for the encouraging words @animeshsingh!

I just made all the changes needed so that the app is now able to work both as a standalone app and in conjunction with Kubeflow. When deployed with Kubeflow it will respect the multi-user guidelines of Kubeflow and perform authorization checks using `SubjectAccessReviews` https://github.com/kubeflow/community/pull/334. The user will also be able to switch between the two modes using an environment variable on the app's manifests.

I also extended the [design doc](https://docs.google.com/document/d/1MaCqHeqd7NiP1nseWkzqqAntRr3PiEH7k_7WnKjQSPU/edit#) to write down the technical details of this as well as how the frontend and the backend will be handling the two modes. This is the same approach we've taken for the Katib UI https://github.com/kubeflow/katib/pull/1483 as well as the other Kubeflow web apps https://github.com/kubeflow/kubeflow/issues/5566

> Also one question I have is that if all the components use the same backend infra for web ui, wouldn't it make sense to have a common dashboard project to hold all the UI code for different Kubeflow components?

Currently, the common code resides on `kubeflow/kubeflow` but the app-specific code should reside on each app's repo. This is what we did for Katib as well https://github.com/kubeflow/katib/pull/1427. So, yes, it may make sense to move the common code to a separate repo, outside of `kubeflow/kubeflow` in the future, but I think this should be an orthogonal and independent discussion that we can have going forward and discuss the pros and cons. But this PR includes KFServing-specific code only.

> Additionally as discussed, please ensure the monitoring can work with monitoring setup in KFS. If not, please suggest changes which will need to go in monitoring setup.
https://github.com/kubeflow/kfserving/blob/master/docs/samples/metrics-and-monitoring/README.md#install-prometheus

I'm looking into this as we speak to ensure the app should work with the monitoring solution KFServing provides. I'll also create a distinct issue for this to discuss what changes are required and what makes sense to extend, and then move forward with the relevant PR.

**Comment by animeshsingh** [comment]
/lgtm

@yuzisun good to go from my side

thanks @kimwnasptd

**Comment by kimwnasptd** [comment]
Thank you very much @animeshsingh!

Also a note, I just rebased the PR on top of the latest master in order to fix the failing tests.

**Comment by yuzisun** [comment]
thanks @kimwnasptd ! will approve once test passes.

**Comment by HassanOuda** [comment]
Will this be part of 1.3 release?

**Comment by kimwnasptd** [comment]
@yuzisun just a kind ping that the tests are passing

**Review by yuzisun — COMMENTED** [review]
@kimwnasptd I think we need to follow up a few things after this is in
1)  Add github action for building and publishing the web app docker images. See how currently KFServing is doing for other images https://github.com/kubeflow/kfserving/tree/master/.github/workflows
2) Add backend unit tests, e2e tests to ensure compatibility. 
3) Create UI deployment manifests addon and document release process.

Would you please help create the issues to track these? thanks!

**Comment by RFMVasconcelos** [comment]
This really looks great @kimwnasptd ! Thank you for all the amazing work :rocket: 

Let's sync to have this well documented so that users arrive to this beautiful experience :)

File: web-app/backend/apps/common/utils.py
Comment (kimwnasptd) [reply web-app/backend/apps/common/utils.py:50]: That's a way better approach from parsing names!
I'll add a commit to address this

**Comment by kimwnasptd** [comment]
> @kimwnasptd I think we need to follow up a few things after this is in

@yuzisun I'll open the issues now, so that we can start tracking these efforts as well. I might ask some questions on the relevant issue for integrating the app with the existing CI but I believe I have what I need.

> This really looks great @kimwnasptd ! Thank you for all the amazing work 
> Let's sync to have this well documented so that users arrive to this beautiful experience :)

@RFMVasconcelos thank you for the kind words!

I'll make sure to keep you in the loop to ensure we document and make it as easy as possible for people to try out this app and give us feedback.

File: web-app/backend/apps/common/utils.py
Comment (kimwnasptd) [reply web-app/backend/apps/common/utils.py]: @yuzisun I pushed a commit that uses the `serving.knative.dev/revision` label to filter which Pods belong to a specific Revision. I also tested it and ensure everything is working as expected.

I also added a [small section](https://docs.google.com/document/d/1MaCqHeqd7NiP1nseWkzqqAntRr3PiEH7k_7WnKjQSPU/edit#heading=h.rm3tua9urp81) in the design doc to keep track of this decision.

I believe this can be resolved so I'll resolve the convo, but feel free to add any further comments if you'd like some more changes

**Comment by kimwnasptd** [comment]
@yuzisun  I created all the issues to track the next steps and efforts, so I think we should be ready now to merge the current PR.

I will provide the corresponding manifests right after the PR is merged, so we can have the Manifests issue closed as well.

**Comment by yuzisun** [comment]
Thanks @kimwnasptd ! This is really awesome work!

/lgtm
/approve

**Pull request merged**
